### PR TITLE
CLC-4943 - Add-Delete: Refactor front-end to use directive for rendering of section lists

### PR DESCRIPTION
**Pull request by redconfetti**
https://jira.ets.berkeley.edu/jira/browse/CLC-4943

### CODE REVIEW

File: src/assets/templates/canvas_embedded/_shared/official_sections_table.html [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+        </th>
+      </tr>
+    </thead>
+    <tbody data-ng-repeat="section in sectionsList">
+      <tr data-ng-class="{'cc-page-course-official-sections-table-row':((listMode !== 'availableStaging' && !$last) || (listMode === 'availableStaging' && !$last && !section.sites)), 'cc-page-course-official-sections-table-row-last':((listMode !== 'availableStaging' && $last) || (listMode === 'availableStaging') && ($last || section.sites)), 'cc-page-course-official-sections-table-row-added':(listMode === 'currentStaging' && section.stagedState === 'add'), 'cc-page-course-official-sections-table-row-deleted':(listMode === 'availableStaging' && section.stagedState === 'delete'), 'cc-page-course-official-sections-table-row-disabled':(listMode === 'availableStaging' && section.isCourseSection && (section.stagedState !== 'delete') || (section.stagedState === 'add'))}" data-ng-if="(listMode === 'preview') || (listMode === 'availableStaging') || (listMode === 'currentStaging' && section.isCourseSection && section.stagedState !== 'delete') || (listMode === 'currentStaging' && !section.isCourseSection && section.stagedState === 'add')">
```
Comment (christianv) [inline src/assets/templates/canvas_embedded/_shared/official_sections_table.html]: @redconfetti could you test whether everything still works if we split the `ng-class` up on multiple lines to make it more readable? If not, leaving it as is should be fine.

File: src/assets/templates/canvas_embedded/_shared/official_sections_table.html [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+          </div>
+        </td>
+      </tr>
+    </tbody>
+    <tbody data-ng-if="listMode === 'currentStaging' && noCurrentSections()">
+      <tr>
+        <td colspan="7">No sections to remain in course site</td>
+      </tr>
+    </tbody>
+
```
Comment (christianv) [inline src/assets/templates/canvas_embedded/_shared/official_sections_table.html]: @redconfetti remove extra line

File: src/assets/templates/canvas_embedded/course_manage_official_sections.html [hunk 3]
```diff
@@ -112,48 +79,7 @@ <h3 class="cc-page-site-provisioning-header cc-page-course-official-sections-exi
...
-                          </div>
-                        </td>
-                      </tr>
-                      <tr data-ng-if="noCurrentSections()">
-                        <td colspan="7">No sections to remain in course site</td>
-                      </tr>
-                    </tbody>
-                  </table>
-                </div>
+                <official-sections-table data-list-mode="'currentStaging'" data-sections-list="allSections" data-unstage-action="unstage(section)" data-stage-delete-action="stageDelete(section)" data-no-current-sections="noCurrentSections()"></official-sections-table>
```
Comment (christianv) [inline src/assets/templates/canvas_embedded/course_manage_official_sections.html]: @redconfetti can you use a `div` as the directive and not a custom element? (For all instances)

```
<div data-bc-official-sections-table ...
```

File: src/assets/javascripts/angular/directives/officialSectionsTableDirective.js [hunk 4]
```diff
@@ -0,0 +1,18 @@
+(function(angular) {
+  'use strict';
+
+  angular.module('calcentral.directives').directive('officialSectionsTable', function() {
```
Comment (christianv) [inline src/assets/javascripts/angular/directives/officialSectionsTableDirective.js]: @redconfetti can we namespace the directive? `bcOfficialSectionsTable` so we know it's related to bCourses

File: src/assets/javascripts/angular/directives/officialSectionsTableDirective.js [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+    return {
+      restrict: 'E',
+      templateUrl: 'canvas_embedded/_shared/official_sections_table.html', // markup for template
```
Comment (christianv) [inline src/assets/javascripts/angular/directives/officialSectionsTableDirective.js]: @redconfetti Can you change `markup` to `Markup`

File: src/assets/javascripts/angular/directives/officialSectionsTableDirective.js [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+      scope: {
+        sectionsList: '=',      // Attribute used to pass an array of sections to render via the template: data-sections-list="sections"
```
Comment (christianv) [inline src/assets/javascripts/angular/directives/officialSectionsTableDirective.js]: @redconfetti Can you remove the extra spaces after the `,` for consistency reasons in our code?

**Comment by christianv** [comment]
@redconfetti Thank you so much for making this change, looks way cleaner now. Just a couple of small comments

File: src/assets/templates/canvas_embedded/course_manage_official_sections.html (hunk 3 above)
Comment (redconfetti) [inline src/assets/templates/canvas_embedded/course_manage_official_sections.html]: Nope. Tried it and it broke it.

File: src/assets/javascripts/angular/directives/officialSectionsTableDirective.js (hunk 4 above)
Comment (redconfetti) [inline src/assets/javascripts/angular/directives/officialSectionsTableDirective.js]: @christianv Yep Did this.

File: src/assets/javascripts/angular/controllers/pages/canvasCourseManageOfficialSectionsController.js [hunk 7]
```diff
@@ -283,6 +283,32 @@
     };
 
     /*
+     * Returns classes to be applied to section row.
+     * Default table row style with bottom border only applied when not the last row displayed, or when an associated sites row is not present.
+     */
+    $scope.rowClassLogic = function(listMode, section, last) {
+      return {
+        'cc-page-course-official-sections-table-row': (!last && !section.sites),
+        'cc-page-course-official-sections-table-row-last': ((listMode !== 'availableStaging' && last) || (listMode === 'availableStaging') && (last || section.sites)),
```
Comment (nicolaasmatthijs) [inline src/assets/javascripts/angular/controllers/pages/canvasCourseManageOfficialSectionsController.js]: I find this type of logic to be really dense to approach, so my preference would be to try to split it out where possible and add sufficient docs

File: src/assets/javascripts/angular/controllers/pages/canvasCourseManageOfficialSectionsController.js [hunk 8]
```diff
@@ -283,6 +283,32 @@
     };
 
     /*
+     * Returns classes to be applied to section row.
+     * Default table row style with bottom border only applied when not the last row displayed, or when an associated sites row is not present.
```
Comment (nicolaasmatthijs) [inline src/assets/javascripts/angular/controllers/pages/canvasCourseManageOfficialSectionsController.js]: Is there a convention that function parameters should be documented?

File: src/assets/javascripts/angular/directives/bcOfficialSectionsTableDirective.js [hunk 9]
```diff
@@ -0,0 +1,20 @@
+(function(angular) {
+  'use strict';
+
+  angular.module('calcentral.directives').directive('bcOfficialSectionsTable', function() {
+    return {
+      restrict: 'AE',
```
Comment (nicolaasmatthijs) [inline src/assets/javascripts/angular/directives/bcOfficialSectionsTableDirective.js]: Given that this is now only being used as an attribute directive, can this change to `restrict: 'A',` ?

**Comment by nicolaasmatthijs** [comment]
Added some additional feedback. Back to @redconfetti

**Comment by nicolaasmatthijs** [comment]
This looks good to me. Merged.

**Pull request merged**